arch/arc/core/smp.c

```diff
@@ -130,21 +131,27 @@ static void sched_ipi_handler(const void *unused)
 	z_sched_ipi();
 }
 
-/* arch implementation of sched_ipi */
-void arch_sched_ipi(void)
+void arch_sched_directed_ipi(uint32_t cpu_bitmap)
 {
-	uint32_t i;
+	unsigned int i;
+	unsigned int num_cpus = arch_num_cpus();
 
-	/* broadcast sched_ipi request to other cores
+	/* Send sched_ipi request to other cores
 	 * if the target is current core, hardware will ignore it
 	 */
-	unsigned int num_cpus = arch_num_cpus();
 
 	for (i = 0U; i < num_cpus; i++) {
-		z_arc_connect_ici_generate(i);
+		if ((cpu_bitmap & BIT(i)) != 0) {
+			z_arc_connect_ici_generate(i);
+		}
 	}
 }
 
+void arch_sched_broadcast_ipi(void)
```

Comment:
arch_sched_broadcast_ipi() I don't think should be using ALWAYS_INLINE as it is not being called from anywhere else in this file. Although arch_sched_directed_ipi() could, I don't think it would get us too much--probably better to let the compiler determine if arch_sched_broadcast_ipi() inlines arch_sched_directed_ipi() or makes a call out to it.